synapse/appservice/api.py

```diff
@@ -444,6 +444,44 @@ async def claim_client_keys(
 
         return response, missing
 
+    async def query_keys(
+        self, service: "ApplicationService", query: Dict[str, List[str]]
+    ) -> Dict[str, Dict[str, Dict[str, JsonDict]]]:
+        """Query the application service for keys.
+
+        Args:
+            query: An iterable of tuples of (user ID, device ID, algorithm).
+
+        Returns:
+            A map of device_keys/master_keys/self_signing_keys/user_signing_keys:
+
+            device_keys is a map of user ID -> a map device ID -> device info.
+        """
+        if service.url is None:
+            return {}
+
+        # This is required by the configuration.
+        assert service.hs_token is not None
+
+        uri = f"{service.url}/_matrix/app/unstable/org.matrix.msc3984/keys/query"
+        try:
+            response = await self.post_json_get_json(
+                uri,
+                query,
+                headers={"Authorization": [f"Bearer {service.hs_token}"]},
+            )
+        except CodeMessageException as e:
+            # The appservice doesn't support this endpoint.
+            if e.code == 404 or e.code == 405:
```

Comment:
not clear from MSC: are we definitely sure this means 'not implemented' and not 'not found'?

Reply:
(likely this one means the MSC should talk about error conditions :/)

Reply:
These are standard error responses used throughout Matrix: https://spec.matrix.org/v1.6/application-service-api/#unknown-routes

Reply:
(Yes, it should likely check the `errcode`, but I'm being a bit more lenient since that behavior is somewhat new.)

Reply:
Well, this is sort of what I mean — 404 without errcode is ambiguous and is non-standard to mean 'not implemented'. It is worth noting that some other endpoints already use 404s for 'not found', so it's not just academic.
Probably we should check the errcode so this is actually following the standard...

(before it's too late and we *do* need 'not found')

Reply:
OK. It is really just disabling the logging in that case anyway, so I guess if they don't return the error code it would just be some additional warning logs.

Reply:
I pulled out the code we use for federation for checking for unknown endpoints, it does a few more checks than just the "correct" ones, but I'd rather not implement a bunch of different functions with logic for finding unknown endpoints.

Reply:
think that looks OK now, I think — thanks!